### ISSUE TITLE
dataProvider cannot have sync methods due to bug in useDataProvider

### ISSUE THREAD

**Issue by deep-choice** [marmelab/react-admin#9887]
**What you were expecting:**
In addition to the prescribed methods like `getOne`, it should be allowed that a dataProvider can have any additional methods. This is probably already the case, but sync methods do not work.

**What happened instead:**
When I do `const dataProvider = useDataProvider()` and the dataProvider has any sync method, an error is thrown.

**Steps to reproduce:**
1. Create a dataProvider with a sync method, such as `getNativeClient()` to expose the native client that actually provides the data.
2. Do `const dataProvider = useDataProvider()` in any component.

**Related code:**
The reason for the error is probably that in `useDataProvider`, **all** methods of the dataProvider are proxied, and the proxy does the following:

```
    const dataProviderProxy = useMemo(() => {
        return new Proxy(dataProvider, {
            get: (target, name) => {
                if (typeof name === 'symbol' || name === 'then') {
                    return;
                }
                return (...args) => {
                    const type = name.toString();

                    if (typeof dataProvider[type] !== 'function') {
                        throw new Error(
                            `Unknown dataProvider function: ${type}`
                        );
                    }

                    try {
                        return dataProvider[type]
                            .apply(dataProvider, args)
                            .then(response => { // ERROR here because the return value of a sync method does not have .then
```

**Environment**

* React-admin version: 4.16.9
* Last version that did not exhibit the issue (if applicable): -
* React version: 18.2.0
* Browser: Firefox
* Stack trace (in case of a JS error): 'The dataProvider threw an error. It should return a rejected Promise instead.'

**Comment by fzaninotto** [comment]
Right, but the data provider is reserved for communication with the API, which is always async. Can you explain the use case you have that requires a sync method? Why can'y you make it async?

**Comment by deep-choice** [comment]
In my case, I want to expose the underlying native API client for more complex requests, so that I do not need to replicate them in the data provider. E.g., I am using Supabase, and the client supports a fluent interface, which would be effortful to repeat.

Besides, there are other valid cases for sync functions, such as simple getters, e.g. for the API credentials.

These functions could be made async theoretically, but this for one thing is not a good architecture because they are not async by their nature, and more importantly, if I do so, I get unnecessary complexity in the React components. An example:

With sync function:
````
const MyComponent = (props: any) => {
  const dataProvider = useDataProvider();
  const refresh = useRefresh();
  const nativeClient = dataProvider.getNativeClient();
  nativeClient.channel('messages').subscribe(() => {refresh();}); // refresh on each message
};
````

With artificially async function:
````
const MyComponent = (props: any) => {
  const dataProvider = useDataProvider();
  const refresh = useRefresh();

  useEffect(() => {
    (async () => {
      const nativeClient = await dataProvider.getNativeClient();>
       nativeClient.channel('messages').subscribe(() => {refresh();}); // refresh on each message
     })();
  }, []);
}
````

Also, when we look at the code of `useDataProvider` again, note that in the proxy the method name (called "type" here) is checked multiple times. A solution idea may be to perform this check outside of the proxy and invoke the proxy only for the prescribed methods such as `getOne`. Alternatively, invoke it only for async methods.

**Comment by djhi** [comment]
Why not just export the supabase client from a file and import it from there?

**Comment by fzaninotto** [comment]
I agree with @djhi. What you describe isn't a use case for which the data provider was designed or that we want to support. So I'm closing this issue as we won't "fix" it. 

I notice that you want to use the data provider to subscribe to a topic. React-admin supports built-in `publish` and `subscribe` methods in the data provider via the realtime module ([see realtime doc](https://marmelab.com/react-admin/Realtime.html)).
